src/object/cli_ec.c

```diff
@@ -0,0 +1,443 @@
+
+	D_ALLOC_ARRAY(params->nsgl.sg_iovs, (params->p_segs.p_nr + sgl->sg_nr));
+	if (params->nsgl.sg_iovs == NULL)
+		return -DER_NOMEM;
+	for (i = 0; i < params->p_segs.p_nr; i++) {
+		params->nsgl.sg_iovs[i].iov_buf = params->p_segs.p_bufs[i];
+		params->nsgl.sg_iovs[i].iov_buf_len = len;
+		params->nsgl.sg_iovs[i].iov_len = len;
+		params->nsgl.sg_nr++;
+	}
+	for (i = 0; i < sgl->sg_nr; i++)
+		params->nsgl.sg_iovs[params->nsgl.sg_nr++] = sgl->sg_iovs[i];
+
+	return rc;
+}
+
+/* Recover EC allocated memory */
+static void
+ec_free_params(struct ec_params *head)
+{
+	D_FREE(head->iods);
+	D_FREE(head->sgls);
+	while (head != NULL) {
+		int i;
+		struct ec_params *current = head;
+
+		D_FREE(current->niod.iod_recxs);
+		D_FREE(current->nsgl.sg_iovs);
+		for (i = 0; i < current->p_segs.p_nr; i++)
+			D_FREE(current->p_segs.p_bufs[i]);
+		D_FREE(current->p_segs.p_bufs);
+		head = current->next;
+		D_FREE(current);
+	}
+}
+
+
+/* Call-back that recovers EC allocated memory  */
+static int
+ec_free_params_cb(tse_task_t *task, void *data)
+{
+	struct ec_params *head = *((struct ec_params **)data);
+	int		  rc = task->dt_result;
+
+	ec_free_params(head);
+	return rc;
+}
+
+/* Identifies the applicable subset of forwarding targets for non-full-stripe
+ * EC updates.
+ */
+void
+ec_init_tgt_set(daos_iod_t *iods, unsigned int nr,
+		struct daos_oclass_attr *oca, unsigned long *tgt_set)
+{
+	unsigned int    len = oca->u.ec.e_len;
+	unsigned int    k = oca->u.ec.e_k;
+	unsigned int    p = oca->u.ec.e_p;
+	unsigned long	par_only = (1UL << p) - 1;
+	unsigned long	full = (1UL <<  (k+p)) - 1;
+	unsigned int	i, j;
+
+	for (i = 0; i < p; i++)
+		*tgt_set |= 1UL << i;
+	for (i = 0; i < nr; i++) {
+		if (iods->iod_type != DAOS_IOD_ARRAY)
+			continue;
+		for (j = 0; j < iods[i].iod_nr; j++) {
+			unsigned long ext_idx;
+			unsigned long rs = iods[i].iod_recxs[j].rx_idx *
+						iods[i].iod_size;
+			unsigned long re = iods[i].iod_recxs[j].rx_nr *
+						iods[i].iod_size + rs - 1;
+
+			/* No partial-parity updates, so this function won't be
+			 * called if parity is present.
+			 */
+			D_ASSERT(!(PARITY_INDICATOR & rs));
+			for (ext_idx = rs; ext_idx <= re; ext_idx += len) {
+				unsigned int cell = ext_idx/len;
+
+				*tgt_set |= 1UL << (cell+p);
+				if (*tgt_set == full) {
+					*tgt_set = 0;
+					return;
+				}
+			}
+		}
+	}
+	/* In case it's a single value, set the tgt_set to send to all.
+	 * These go everywhere for now.
+	 */
+	if (*tgt_set == par_only)
+		*tgt_set = 0;
+}
+
+/* Iterates over the IODs in the update, encoding all full stripes contained
+ * within each recx.
+ */
+int
+ec_obj_update(tse_task_t *task, daos_obj_id_t oid, daos_oclass_attr_t *oca,
+	      uint64_t *tgt_set)
+{
+	daos_obj_update_t	*args = dc_task_get_args(task);
+	struct ec_params	*head = NULL;
+	struct ec_params	*current = NULL;
+	unsigned int		 i, j;
+	int			 rc = 0;
+
+	for (i = 0; i < args->nr; i++) {
+		if (ec_has_full_stripe(&(args->iods[i]), oca, tgt_set)) {
```

Comment:
style: how about declaring local variables like:
```
daos_sg_list_t *sgl = &args->sgls[i];
daos_iod_t       *iod = &args->iods[i];
```
By doing this, this code block can be easier to read?